### PR TITLE
docs: thomas/update mw docs

### PR DESCRIPTION
**Pull request by tmsdkeys**
## Description

Updated the middleware documentation after receiving feedback:

- [X] Split files into overview, mw dev and chain dev sections
- [X] Added middleware diagram
- [X] Added more info on `IBCMiddleware` struct
- [X] Clarified ICS4Wrapper concept in `develop.md`
- [X] Added a note on difference between stateless and stateful mw to access `ics4Wrapper`
- [X] Updated links to v7 code, updated some terminology
- [X] Updated `integration.md` for chain devs

### Commit Message / Changelog Entry

```bash
docs: update middleware docs
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [X] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [X] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/10-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [X] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [X] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Comment by tmsdkeys** [comment]
Last bit of work to finish this draft PR would be to take a look at the `integration.md` section. There we have some middlewares, like `mw3` appearing in multiple stacks:

```
// create IBC stacks by combining middleware with base application
// NOTE: since middleware2 is stateless it does not require a Keeper
// stack 1 contains mw1 -> mw3 -> transfer
// stack 2 contains mw3 -> mw2 -> custom1
```
so `mw3` here is being used in 2 stacks, and afaiu we'd need two different keepers:

```go
mw3Keeper1 := mw3.NewKeeper(..., ics4Wrapper: mw1Keeper, ...) //  in stack 1
mw3Keeper2 := mw3.NewKeeper(..., ics4Wrapper: channelKeeper, ...) //  in stack 2
```

rather than simply:
```go
mw3Keeper := mw3.NewKeeper(storeKey3)
```

And then defining the stacks as:
```diff
- stack1 := mw1.NewIBCMiddleware(mw3.NewIBCMiddleware(transferIBCModule, mw3Keeper), mw1Keeper)
+ stack1 := mw1.NewIBCMiddleware(mw3.NewIBCMiddleware(transferIBCModule, mw3Keeper1), mw1Keeper)

- stack2 := mw3.NewIBCMiddleware(mw2.NewIBCMiddleware(customIBCModule1), mw3Keeper)
+ stack2 := mw3.NewIBCMiddleware(mw2.NewIBCMiddleware(customIBCModule1), mw3Keeper2)
```

Similarly, should we not also register them as different SDK modules in the module manager?
```diff
// Only create App Module **once** and register in app module
// if the module maintains independent state and/or processes sdk.Msgs
app.moduleManager = module.NewManager(
  ...
  mw1.NewAppModule(mw1Keeper),
-   mw3.NewAppModule(mw3Keeper),
+   mw3.NewAppModule(mw3Keeper1),
+   mw3.NewAppModule(mw3Keeper2),
  transfer.NewAppModule(transferKeeper),
  custom.NewAppModule(customKeeper)
)
```

File: docs/ibc/middleware/integration.md [hunk 1]
```diff
@@ -19,23 +19,22 @@ The order of middleware **matters**, function calls from IBC to the application
...
   custom.NewAppModule(customKeeper)
 )
 
-mw1IBCModule := mw1.NewIBCModule(mw1Keeper)
-mw2IBCModule := mw2.NewIBCModule() // middleware2 is stateless middleware
-mw3IBCModule := mw3.NewIBCModule(mw3Keeper)
-
 scopedKeeperTransfer := capabilityKeeper.NewScopedKeeper("transfer")
 scopedKeeperCustom1 := capabilityKeeper.NewScopedKeeper("custom1")
 scopedKeeperCustom2 := capabilityKeeper.NewScopedKeeper("custom2")
```
Comment (crodriguezvega) [inline docs/ibc/middleware/integration.md:40]: These scoped keepers are created here, but not used anywhere else... Should we show pseudocode as well for the keepers that use them?

**Review by crodriguezvega — COMMENTED** [review]
Thank for this nice refactoring of the docs, @tmsdkeys!

File: docs/ibc/middleware/integration.md (hunk 1 above)
Comment (tmsdkeys) [reply docs/ibc/middleware/integration.md:40]: I had not paid too much attention here to the scoped keepers but I think you're correct

Comment (tmsdkeys) [reply docs/ibc/middleware/integration.md:40]: Updated this

**Review by damiannolan — APPROVED** [review]
Nice one @tmsdkeys! Sorry about delayed review, honestly thought I had already approved this 😆 

LGTM 🚀 🚀 🚀

**Pull request merged**